graders/c/cgrader/cgrader.py

```diff
@@ -302,69 +301,90 @@ def test_run(
 
         out = self.run_command(
             command if args is None else ([command] + args),
             input,
             sandboxed=True,
             timeout=timeout,
         )
+
         outcmp = out
+        if highlight_matches and out:
+            for _, r in exp_output:
+                out = r.sub(f"\\033[32m\\g<0>\\033[0m", out)
+            for _, r in reject_output:
+                out = r.sub(f"\\033[31m\\g<0>\\033[0m", out)
         if not out:
             out = "(NO OUTPUT)"
         elif not out.endswith("\n"):
             out += "\n(NO ENDING LINE BREAK)"
 
-        if ignore_case:
-            outcmp = outcmp.lower()
-            if exp_output:
-                exp_output = [str(t).lower() for t in exp_output]
-            if reject_output:
-                reject_output = [str(t).lower() for t in reject_output]
-
-        if ignore_consec_spaces:
-            # Replace all space-like characters with single space
-            outcmp = re.sub(r"\s+", " ", outcmp)
-            if exp_output:
-                exp_output = [re.sub(r"\s+", " ", str(t)) for t in exp_output]
+        if msg is None and exp_output:
+            comment = (
+                ""
+                if len(exp_output) == 1
+                else " all of"
+                if must_match_all_outputs
+                else " one of"
+            )
+            msg = f"Expected{comment}:\n\t" + "\n\t".join(
+                [
+                    f"\033[32m{t}\033[0m"
+                    if highlight_matches and r.search(outcmp)
+                    else t
+                    for t, r in exp_output
+                ]
+            )
             if reject_output:
-                reject_output = [re.sub(r"\s+", " ", str(t)) for t in reject_output]
+                msg += "\nBut not:\n\t" + "\n\t".join(
+                    [
+                        f"\033[31m{t}\033[0m"
+                        if highlight_matches and r.search(outcmp)
+                        else t
+                        for t, r in reject_output
+                    ]
+                )
+        elif msg is None:
+            msg = ""
 
         points = True
-        if timeout and "TIMEOUT" in out:
+        if timeout and "TIMEOUT" in outcmp:
             points = False
-        elif size_limit and len(out) > size_limit:
+        elif size_limit and len(outcmp) > size_limit:
             out = out[0:size_limit] + "\nTRUNCATED: Output too long."
             points = False
-        elif (
-            exp_output is not None
-            and must_match_all_outputs
-            and [t for t in exp_output if str(t) not in outcmp]
-        ):
-            points = False
-        elif (
-            exp_output is not None
-            and not must_match_all_outputs
-            and not [t for t in exp_output if str(t) in outcmp]
+        elif not (all if must_match_all_outputs else any)(
+            [r.search(outcmp) for _, r in exp_output]
```

Comment:
I don't think you need to have these in a list, `all` and `any` can take in arbitrary generators. Also, better to be explicit about checking for `None` vs a match object.
```suggestion
            r.search(outcmp) is not None for _, r in exp_output
```

Reply:
Resolved in ea794457e.